fendermint/vm/interpreter/src/genesis.rs

```diff
@@ -0,0 +1,765 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::collections::{BTreeSet, HashMap};
+use std::io::{Read, Write};
+use std::marker::PhantomData;
+use std::path::{Path, PathBuf};
+use std::pin::Pin;
+use std::sync::Arc;
+
+use anyhow::{anyhow, Context};
+use base64::Engine;
+use cid::Cid;
+use ethers::abi::Tokenize;
+use ethers::core::types as et;
+use fendermint_actor_eam::PermissionModeParams;
+use fendermint_eth_hardhat::{ContractSourceAndName, Hardhat, FQN};
+use fendermint_vm_actor_interface::diamond::{EthContract, EthContractMap};
+use fendermint_vm_actor_interface::eam::EthAddress;
+use fendermint_vm_actor_interface::ipc::IPC_CONTRACTS;
+use fendermint_vm_actor_interface::{
+    account, burntfunds, chainmetadata, cron, eam, init, ipc, reward, system, EMPTY_ARR,
+};
+use fendermint_vm_core::{chainid, Timestamp};
+use fendermint_vm_genesis::{ActorMeta, Collateral, Genesis, Power, PowerScale, Validator};
+use futures_util::io::Cursor;
+use fvm::engine::MultiEngine;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_car::{load_car, CarHeader};
+use fvm_ipld_encoding::CborStore;
+use fvm_shared::chainid::ChainID;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::version::NetworkVersion;
+use ipc_actors_abis::i_diamond::FacetCut;
+use num_traits::Zero;
+
+use crate::fvm::state::snapshot::{derive_cid, StateTreeStreamer};
+use crate::fvm::state::{FvmGenesisState, FvmStateParams};
+use crate::fvm::store::memory::MemoryBlockstore;
+use fendermint_vm_genesis::ipc::IpcParams;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use tokio_stream::StreamExt;
+use tokio_util::compat::TokioAsyncWriteCompatExt;
+
+/// The sealed genesis state metadata
+#[serde_as]
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
+struct GenesisMetadata {
+    pub state_params: FvmStateParams,
+    pub validators: Vec<Validator<Power>>,
+}
+
+impl GenesisMetadata {
+    fn new(state_root: Cid, out: GenesisOutput) -> GenesisMetadata {
+        let state_params = FvmStateParams {
+            state_root,
+            timestamp: out.timestamp,
+            network_version: out.network_version,
+            base_fee: out.base_fee,
+            circ_supply: out.circ_supply,
+            chain_id: out.chain_id.into(),
+            power_scale: out.power_scale,
+            app_version: 0,
+        };
+
+        GenesisMetadata {
+            state_params,
+            validators: out.validators,
+        }
+    }
+}
+
+/// Genesis app state wrapper for cometbft
+#[repr(u8)]
+pub enum GenesisAppState {
+    V1(Vec<u8>) = 1,
+}
+
+impl GenesisAppState {
+    pub fn v1(bytes: Vec<u8>) -> Self {
+        Self::V1(bytes)
+    }
+
+    pub fn compress_and_encode(&self) -> anyhow::Result<String> {
+        let bytes = match self {
+            GenesisAppState::V1(ref bytes) => {
+                let mut wtr = snap::write::FrameEncoder::new(vec![0]);
```

Comment:
@cryptoAtwill this looks like a bug? We write the version discriminator (0) under snappy compression, but we expect to find it uncompressed below?